### DISCUSSION TITLE
Propagating cancellation to an enqueued task

### DISCUSSION THREAD

**Discussion by Cykelero** [dimasusername/Queue#3] [Q&A]
I'm trying to wrap my head around this.

From some async code, I'm enqueing a task onto a queue:
```swift
// Outer task
let task = queue.addOperation {
	// Inner task
	doFirstThing()
	guard !Task.isCancelled else { return }
	doSecondThing()
}

return await task.value
```

If the outer task gets cancelled, the inner task doesn't get cancelled too—there's no parent-child relationship.

Would propagating cancellation be achievable?
And would it be a good idea at all? (i.e. does my need to propagate cancellation suggest I'd be better served by something other than a queue)

The queue is used to make sure a given Web resource is only being written to by a single operation at a time. Or, in other words, my enqueing is not fire-and-forget; the tasks are kept track of, surfaced in the app's UI, and the user can elect to cancel them.

**Comment by mattmassicotte** [comment]
Hmmm. There is a test that attempts to validate this exact thing, and it passes. These kinds of things can be racy though. Could it be the test is wrong? Or perhaps your code is actually being canceled *after* the check is performed?

https://github.com/mattmassicotte/Queue/blob/cc86ecf909e1a8e724c56c488b70d4d476d73ca1/Tests/QueueTests/AsyncQueueTests.swift#L209

**Reply by Cykelero** [reply]
Doesn't the test check that you can cancel the _inner_ task?
What I'm trying to achieve is propagating a cancellation of the outer task, the the inner one. E.g.:
```swift
let outerTask = Task {
	let innerTask = queue.addOperation {
		try await Task.sleep(for: .seconds(3))
		print("Inner task was not canceled")
		
		return try await doSomeWork()
	}
	
	try await Task.sleep(for: .seconds(2))
	print("Outer task was not canceled")
	
	return try await innerTask.value
}

try await Task.sleep(for: .seconds(1))

outerTask.cancel()
```
This prints “Inner task was not canceled”.

**Reply by Cykelero** [reply]
So, this seems to work:
```swift
let task = queue.addOperation {
	return try await doSomeWork()
}

return try await withTaskCancellationHandler {
	try await task.value
} onCancel: {
	task.cancel()
}
```

I'm still very curious to get your opinion on this—whether or not it's likely to be a good idea.

The code in that last return could easily be wrapped in some utility `propagatingCancellation(task: Task)` method, too.

**Reply by mattmassicotte** [reply]
Ohhh, I misunderstood that you had wrapped all this up in another Task! Yes using `withTaskCancellationHandler` can do this.

But, I think you should be thinking really hard about using a pattern like this. Your outer Task no longer has ordering with respect to its caller. Are you *certain* this is not introducing races?

**Comment by Cykelero** [comment]
I _think_ it is. The outer task is started from a UI callback (e.g. when the user clicks the Push button), so its resolution doesn't have a strong ordering requirement.

The goal of that outer `Task {}` is to keep track of the long-running operation, both to display its progress, and allowing it to be cancelled by the user:
![2024-02-01 19 15 30 Retcon (Debug) app (test-DarkModeBuddy)](https://github.com/mattmassicotte/Queue/assets/2979318/83c30936-db1f-427f-955b-9bdfe0d6fa78)

So if the outer task is cancelled, the inner task really needs to be cancelled to—regardless of whether it's waiting for its turn in a queue, or not.

**Reply by mattmassicotte** [reply]
What happens if the user pressed the button and *before the outer task is executed* the user presses another button that accesses that queue?

**Reply by Cykelero** [reply]
Sorry, I don't understand. By “before the outer task is executed”, do you mean before it's started, or ended?
Either way I can't think of what sort of issue could result. What do you have in mind?

**Reply by mattmassicotte** [reply]
Before it is started.

Button press 1:

```swift
let outerTask1 = Task { print("1" }
```

Then later, button press 2:

```swift
let outerTask2 = Task { print("2" }
```

This can result in the output:

```
2
1
```

In your specific usage it may not matter though!

**Reply by Cykelero** [reply]
Ah, I think I get it! If the app's threads are already busy, the first press' outer task could get delayed to after the second button press, and so get ordered incorrectly.

I'll probably leave things like this for now—that situation wouldn't be too problematic—but I'll keep an eye out for this sort of behavior. Thank you for the explanations!

**Reply by mattmassicotte** [reply]
Even if the threads are not busy that could happen! Task is entirely unordered in every way. I'm slightly surprised this is ok but you also need a queue, but I totally believe it could be true.

And you are very welcome!
